### PR TITLE
Added support for running the API in container

### PR DESCRIPTION
**Pull request by pushrbx**
Resolves: #261

I've added support for [RoadRunner](https://roadrunner.dev), which is an application server for PHP written in Go. This will allow for running the API in a container, also some self-hosters will be able to run the app in clusters more easily. However I think we will only support the docker/podman approach with our documentation. 

This PR also introduces some changes in the code:
- Changed some bits to make it compliant with php 8.1 as in one of the container images we use that version
- Added some code for roadrunner: PSR error handling and behaviour branching for each supported search engine in case of service warming (on the level of service bindings)
- Added `docker-entrypoint.php` which sets safe defaults, so you don't have to type too many things in the console when you are trying to start a new container.

Also this whole thing relies on a modified `roadrunner-laravel` library, which I've modified to work with Lumen:
- https://github.com/pushrbx/lumen-roadrunner/
- https://github.com/spiral/roadrunner-laravel

Composer will install the `lumen-roadrunner` package from GIT in this case, not sure if I should release it to the wild yet.

TODO:     

- [x] Add github action for distributing the container images

### CODE REVIEW

File: .rr.yaml [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+
+  # Response status code if a requested plugin not ready to handle requests
+  # Valid for both /health and /ready endpoints
+  #
+  # Default: 503
+  unavailable_status_code: 503
+
+service:
+  laravel_queue_worker_1:
+    command: "php /app/artisan queue:work --queue=high,low"
```
Comment (irfan-dahir) [inline .rr.yaml:203]: This has been changed to the following by laravel:
`php artisan queue:work --queue=high,default`

Ref: https://laravel.com/docs/9.x/queues#connections-vs-queues

File: app/Listeners/PsrWorkerErrorListener.php [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    public function __construct()
+    {
+        $this->exceptionHandler = new \App\Exceptions\Handler;
+        $this->logger = new Logger('psr-worker');
+        $this->logger->pushHandler(new StreamHandler(storage_path().'/logs/psr-worker.log'), env('APP_DEBUG') ? Logger::DEBUG : Logger::WARNING);
+    }
+
+    public function handle(LoopErrorOccurredEvent $event): void
+    {
+        $this->logger->info("error logging...");
```
Comment (irfan-dahir) [inline app/Listeners/PsrWorkerErrorListener.php:23]: Intentionally left in?

**Review by irfan-dahir — CHANGES_REQUESTED** [review]
Left some comments. The rest LGTM.

File: app/Listeners/PsrWorkerErrorListener.php (hunk 2 above)
Comment (pushrbx) [reply app/Listeners/PsrWorkerErrorListener.php:23]: copy pasted example code and accidentally left it in, not needed

**Comment by pushrbx** [comment]
One discussion point regarding the container image distribution:
Do we want to release different versions for  each version of php or should we just stick with the latest php version?

**Comment by irfan-dahir** [comment]
@pushrbx I think we should stick to the version we're currently running things on plus the next future release (latest PHP ver. e.g 8.0 and 8.1 in this case). Fewer things to support and look at IMO instead of each ver.

**Comment by pushrbx** [comment]
@irfan-dahir Currently the github action workflow creates two container images: one for php 8.0 and one for php 8.1
Imo from the user perspective they just want to run the container image and that's all, I don't think they would have a preference which php version should be ran. Hence my question.

**Comment by irfan-dahir** [comment]
@pushrbx Hmm, we do have a minimum req. set for [jikan-rest](https://github.com/jikan-me/jikan-rest) and [jikan](https://github.com/jikan-me/jikan) to php 8.0. Won't that cause issues if we bundle for lower PHP vers.?

**Comment by pushrbx** [comment]
I'm not saying that we should bundle lower PHP vers, all I'm saying do we have to bundle for two versions of php in case of container  images 😄  (php 8.0 and php 8.1)

**Comment by pushrbx** [comment]
Btw sorry, I wasn't too specific in my original question. I was referring to the matrix setting in the `container-image-release.yml` file for creating container images: https://github.com/pushrbx/jikan-rest/blob/feature/roadrunner/.github/workflows/container-image-release.yml#L11

**Comment by irfan-dahir** [comment]
Sorry, I don't have much knowledge of containers. I'm assuming they will be prompted for which version to pull?

> Imo from the user perspective they just want to run the container image and that's all, I don't think they would have a preference which php version should be ran.

If this is the case, I think then we can just support for the version we're currently we have requirements on aka. ver 8.0. And then bump that up for 8.1 when we push the req. Wdyt?

**Comment by pushrbx** [comment]
> Sorry, I don't have much knowledge of containers. I'm assuming they will be prompted for which version to pull?
> 
> > Imo from the user perspective they just want to run the container image and that's all, I don't think they would have a preference which php version should be ran.
> 
> If this is the case, I think then we can just support for the version we're currently we have requirements on aka. ver 8.0. And then bump that up for 8.1 when we push the req. Wdyt?

I'm going to remove the matrix thingy then, and we just have to make sure to change the php version in the container image whenever we upgrade the requirement in the composer config.

**Comment by irfan-dahir** [comment]
@pushrbx Sounds good 👌

**Comment by pushrbx** [comment]
Changed the github action for container releases up a bit:
- You need to manually run the base image creation once in a while (this sets up common stuff, and speeds up builds)
- When you create a new release, the github action will run to push a new container image to the registries (DockerHub and GHCR (Packages under org))

We can change the tags a bit later if necessary, like if we push to master there should be an edge version tag.
I've used these guides to improve things (I've learned a lot yeay):

- https://github.com/docker/metadata-action
- https://github.com/orgs/community/discussions/26279
- https://github.com/tradingstrategy-ai/frontend/blob/5fa560dd029a44b5e66df6c63ca2b7884274345a/.github/workflows/javascript.yml#L53
- https://github.com/docker/build-push-action/blob/master/docs/advanced/cache.md#cache-backend-api
- https://github.com/orgs/community/discussions/26686

**Review by irfan-dahir — APPROVED** [review]

**Pull request merged**